x/gamm/keeper/msg_server_test.go

```diff
@@ -279,20 +278,85 @@ func (suite *KeeperTestSuite) TestJoinPool_Events() {
 				suite.Require().NotNil(response)
 			}
 
-			assertEventEmitted(suite, ctx, types.TypeEvtPoolJoined, tc.expectedAddLiquidityEvents)
-			assertEventEmitted(suite, ctx, sdk.EventTypeMessage, tc.expectedMessageEvents)
+			suite.AssertEventEmitted(ctx, types.TypeEvtPoolJoined, tc.expectedAddLiquidityEvents)
+			suite.AssertEventEmitted(ctx, sdk.EventTypeMessage, tc.expectedMessageEvents)
 		})
 	}
 }
 
-func assertEventEmitted(suite *KeeperTestSuite, ctx sdk.Context, eventTypeExpected string, numEventsExpected int) {
-	allEvents := ctx.EventManager().Events()
-	// filter out other events
-	actualEvents := make([]sdk.Event, 0, 1)
-	for _, event := range allEvents {
-		if event.Type == eventTypeExpected {
-			actualEvents = append(actualEvents, event)
-		}
+// TestExitPool_Events tests that events are correctly emitted
+// when calling ExitPool.
+func (suite *KeeperTestSuite) TestExitPool_Events() {
+	const (
+		tokenOutMinAmount = 1
+		shareOut          = 110
+	)
+
+	testcases := map[string]struct {
+		poolId                        uint64
+		shareInAmount                 sdk.Int
+		tokenOutMins                  sdk.Coins
+		expectError                   bool
+		expectedRemoveLiquidityEvents int
+		expectedMessageEvents         int
+	}{
+		"successful exit": {
+			poolId:                        1,
+			shareInAmount:                 sdk.NewInt(shareOut),
+			tokenOutMins:                  sdk.NewCoins(),
+			expectedRemoveLiquidityEvents: 1,
+			expectedMessageEvents:         3, // 1 gamm + 2 tendermint.
```

Comment:
just for my learning, why are there two TM messages emitted here?

Reply:
These are not TM, the comment is wrong, and I'm going to change this. This happens because other keeper methods emit these events. For example, bank send:
https://github.com/cosmos/cosmos-sdk/blob/71035879e4af4551b9bcf16f79bc52785793602e/x/bank/keeper/send.go#L213-L215

I think the bank send is wrong and should instead have its message emitted in the message server so that we don't triple-count it here.